poppy/accel_math.py

```diff
@@ -64,16 +88,213 @@ def _exp(x):
+    #    (x & 31)==0  is a ~20x faster equivalent of (np.mod(x,32)==0)
+    if (_USE_CUDA) & (N==x.shape[1]) & ((N & 31)==0):
         blockdim = (32, 32) # threads per block
         numBlocks = (int(N/blockdim[0]),int(N/blockdim[1]))
         cufftShift_2D_kernel[numBlocks, blockdim](x.ravel(),N)
         return x
     else:
         return np.fft.fftshift(x)
 
+def _ifftshift(x):
+    """ Inverse FFT shifts of array contents, using CUDA if available.
+    Otherwise defaults to numpy.
+    Note, ifftshift and fftshift are identical for even-length x,
+    the functions differ by one sample for odd-length x. This function
+    implictly assumes that if using CUDA, the array size must be even,
+    so we can use the same algorithm as fftshift.
+
+    Note - TODO write an OpenCL version
+
+    See also fftshift
+    """
+
+    N=x.shape[0]
+    # the CUDA fftshift is set up to work on blocks of 32, so
+    # N must be a multiple of 32. We check this rapidly using a bit mask:
+    #   not (x & 31)  is a ~20x faster equivalent of (np.mod(x,32)==0)
+    if (_USE_CUDA) & (N==x.shape[1]) & ((N & 31)==0):
+        blockdim = (32, 32) # threads per block
+        numBlocks = (int(N/blockdim[0]),int(N/blockdim[1]))
+        cufftShift_2D_kernel[numBlocks, blockdim](x.ravel(),N)
+        return x
+    else:
+        return np.fft.ifftshift(x)
+
+
+
+
+def fft_2d(wavefront, forward=True, normalization=None, fftshift=True):
+    """ main entry point for FFTs, used in Wavefront._propagate_fft and
+    elsewhere. This will invoke one of the following, depending on availability:
+        - CUDA on NVidia GPU
+        - OpenCL on AMD GPU
+        - FFTW on CPU
+        - numpy on CPU
+
+    This function handles ONLY the core numerics itself, as fast as possible,
+    (and some minor related logging) .
+    All the interaction with object state for Wavefront arrays should happen elsewhere.
+
+    TODO: this should execute an IN PLACE FFT, so we don't have to pass around arrays to return
+    anything.
+
+    Parameters
+    -----------
+    forward : bool
+        set to True for forward FFT, False for inverse fft
+    normalization : float, optional
+        Normalization factor. Defaults to 1./wavefront.shape[0] for forward,
+        and wavefront.shape[0] for inverse. Use this only if you need a non-default
+        behavior.
+    fftshift : bool
+        apply FFT shift after forwards FFT or before inverse FFT?
+
+    """
+    # To use a fast FFT, it must both be enabled and the library itself has to be present
+    _USE_CUDA = (conf.use_cuda and _CUDA_AVAILABLE)
+    _USE_OPENCL = (conf.use_opencl and _OPENCL_AVAILABLE)
+    _USE_FFTW = (conf.use_fftw and _FFTW_AVAILABLE)
+
+    # OpenCL cfFFT only can FFT certain array sizes.
+    # This check is more stringent that necessary - opencl can handle powers of a few small integers
+    # but this simple version helps during development
+    if _USE_OPENCL and not ispowerof2(wavefront.shape[0]):
```

Comment:
@douglase are there constraints on which array sizes are allowed for CUDA FFTs? How well does it handle odd sized arrays? Not that I expect those to be constant, but we should check for and avoid cases where it would fail. 

FWIW, the clfft library says it supports "powers of 2, 3, 5, 7, 11 and 13". but here I just check for powers of 2, because that's the common case, and I don't know how to write a computationally-efficient fast check for "factorable into powers of primes no greater than 13".  

I don't necessarily put a high priority on optimal GPU acceleration of weird sized arrays, I just want to make sure the code doesn't encounter any situation that causes a fatal error. Although for some situations it _is_ useful to FFT such arrays; for instance if you need bigger than 4096^2, it's still about 2x faster to FFT an array of size 6144^2 than to jump all the way to 8192^2.

Reply:
Ah, this explains why suddenly my odd-sized test was the same size as FFTW. 

CUDA definitely is happy with odd-sizes, but takes a slight hit for unusually sized arrays as seen in this figure from the current master branch where everything is power of two except for a 2800x2800 test:

![benchmarks12coresv100](https://user-images.githubusercontent.com/1025951/40146178-8b40f168-5932-11e8-9441-044c7ce4aee9.png)

This is consistent the cuFFT documentation (https://docs.nvidia.com/cuda/cufft/index.html#introduction):

'''Features:
* Algorithms highly optimized for input sizes that can be written in the form 2^a×3^b×^5^c^×7^d. In general the smaller the prime factor, the better the performance, i.e., powers of two are fastest.
*An O(nlogn)  algorithm for every input data size
...
'''

Reply:
So I would just pass any sized array to CUDA

Reply:
hmm, nope, not using opencl and it's only the opencl case your are testing. maybe it's a coincidence of this system. I will keep digging.

Reply:
Never mind, it was jus a coincidence of that machine/run.  I am seeing a speed up from CUDA at 2800x2800.
![benchmarks12coresv100](https://user-images.githubusercontent.com/1025951/40208603-92b8bd46-5a08-11e8-89f9-49c042e2432d.png)

Reply:
Can you point me at the benchmark code you’re using for that figure?  You pointed me out a notebook a while ago but I don’t know if that’s still the up-to-date place to find it. I’ll make a similar figure for OpenCL. 

Might want to tweak the X axis label to say “ array dimensions per axis”,  to emphasize that the full array size is the square of that.

Reply:
that notebook is still the most current.